pkg/querier/queryrange/queryrangebase/results_cache.go

```diff
@@ -211,7 +211,19 @@ func NewResultsCacheMiddleware(
 	}), nil
 }
 
+// skipRequestType returns whether we should cache the r request type.
+// This is needed when we have middlewares that send different requests types down
+// in the pipeline that do not support caching.
+func (s *resultsCache) skipRequestType(r Request) bool {
```

Comment:
Rather than threading this logic through our middleware chain, what if we had a separate `skip` middleware chain for pre-querying these `IndexStats` requests? Look at how `NewQueryShardMiddleware` embeds two different middleware chains depending on whether we should use sharding or not. This would allow us to handle this without subsequent middlewares needing to be aware of this.

Another idea for later: what if we cached `IndexStats` requests in a results cache? What if we cached other types of index queries in results caches?

Reply:
> Look at how NewQueryShardMiddleware embeds two different middleware chains depending on whether we should use sharding or not. This would allow us to handle this without subsequent middlewares needing to be aware of this.

IIUC, the pattern used in `NewQueryShardMiddleware` is useful only for bypassing that middleware itself, not the middlewares above or below the shard middleware itself in the pipeline.

In other words, when sharding is disabled, the request will skip the sharding-related middleware (`shardingware`) configured in `NewQueryShardMiddleware` but it will still go throughout the rest of middlewares configured in `NewMetricTripperware` after the sharding middleware (for example, it will go throughout the `NewRetryMiddleware`). The same applies to the IndexStats req done at `dynamicShardResolver.Shard` it skips the sharding mechanism, but not whatever is configured after sharding.

Reply:
I've implemented something similar at [648589e](https://github.com/grafana/loki/pull/8670/commits/648589ee3b14af23ec9c6f9a84300f7bec774fb8) and [75b1d54](https://github.com/grafana/loki/pull/8670/commits/75b1d5413cf13d291db4b426ffeb2d62fe419186).